File: Math/273.integer_to_english_words.py
class Solution():
    def one_digit(self, num):
        numbers = [i for i in range(1, 10)]
        words = ["One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine"]
        table = dict(zip(numbers, words))
        return table[num]
    def teen(self, num):
        numbers = [i for i in range(10, 20)]
        words = ["Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"]
        table = dict(zip(numbers, words))
        return table[num]
    def tens(self, num):
        numbers = [i for i in range(2, 10)]
        words = ["Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"]
        table = dict(zip(numbers, words))
        return table[num]
    def three_digits(self, num):
        result = ""
        hundred = num//100
        if hundred:
            result += self.one_digit(hundred) + " Hundred"
        ten = num%100//10
        if ten == 1:
            result += (" " if result else "") + self.teen(num%100)
        else:
            if ten >= 2:
                result += (" " if result else "") + self.tens(ten)
            single = num%10
            if single:
                result += (" " if result else "") + self.one_digit(single)
        return result
    def numberToWords(self, num: int) -> str:
        ans = ""
        billion = num//1000000000
        if billion:
            ans += self.one_digit(billion) + " Billion"
        million = num%1000000000//1000000
        if million:
            ans += (" " if ans else "") + self.three_digits(million) + " Million"
        thousand = num%1000000//1000
        if thousand:
            ans += (" " if ans else "") + self.three_digits(thousand) + " Thousand"
        res = num%1000
        if res:
            ans += (" " if ans else "") + self.three_digits(res)
        if not ans:
            return "Zero"
        else:
            return ans

File: C++/Math/273. Integer to English Words.cpp
class Solution {
public:
    
    vector<string> to19 = {"","One","Two","Three","Four","Five","Six","Seven","Eight","Nine","Ten","Eleven","Twelve","Thirteen","Fourteen","Fifteen","Sixteen","Seventeen","Eighteen","Nineteen"};
    vector<string> tens = {"","","Twenty","Thirty","Forty","Fifty","Sixty","Seventy","Eighty","Ninety"};
    
    vector<string> helper(int n) {
        if (n == 0) {
            return {};
        }
        if (n >= 1 && n < 20) {
            return {to19[n]};
        }
        if (n < 100) {
            vector<string> ans = {tens[n/10]};
            vector<string> remain = helper(n%10);
            ans.insert(ans.end(), remain.begin(), remain.end());
            return ans;
        }
        if (n < 1000) {
            vector<string> ans = {to19[n/100], "Hundred"};
            vector<string> remain = helper(n%100);
            ans.insert(ans.end(), remain.begin(), remain.end());
            return ans;
        }
        vector<pair<int, string>> powers{{1,"Thousand"},{2,"Million"},{3,"Billion"}};
        for (auto [power, word] : powers) { 
            if (n < pow(1000, power+1)) {
                vector<string> ans = helper(n/pow(1000, power));
                ans.push_back(word);
                vector<string> remain = helper(n%(int)pow(1000, power));
                ans.insert(ans.end(), remain.begin(), remain.end());
                return ans;
            }
        }
        return {};
    }
    
    string numberToWords(int num) {
        if (num == 0) return "Zero";
        vector<string> words = helper(num);
        string ans = words[0];
        for (int i = 1; i < words.size(); i++) {
            ans += " " + words[i];
        }
        return ans;
    }
};

